Drop usage of internal API when conditionally including assets

kartben opened the pull request:
Fixes #197 by reversing the logic of how JS/CSS assets are added to pages.

Rather than adding assets to the app and using undocumented API to *remove* them when not needed, do the exact opposite and only *add* them to the local html-page-context when building a page that uses the directive.

kartben commented:
just realized I forgot to update pytests - should be green now

ru-fu reviewed (approved):
My Python skills don't suffice to review the changes in detail, but they look sensible.
I've tested the updated version and everything works as expected, and the warnings are gone.

Thanks a lot!

kartben commented:
@foster999 `codecov/codecov-action@v1` is probably due for an update since v4.5.0 is the current version. Maybe this will help handline rate limit automagically? Do you want me to add a commit taking care of this to this PR?

foster999 commented:
Thanks for this @kartben 😄 

> @foster999 `codecov/codecov-action@v1` is probably due for an update since v4.5.0 is the current version. Maybe this will help handle rate limit and perform retries automagically? Do you want me to add a commit taking care of this to this PR?

Yes please, might as well sort this at the same time! Failing that, I'd be happy to drop the codecov step in the Actions config for now.

Could you also drop the lines in the change for backwards compatibility with old sphinx (and if needed update the setup.py) please? I've tried to move towards only officially supporting the latest major version of dependencies, as older versions of sphinx-tabs can be used with older dependencies

This part of the code has been rewritten to and from this approach a few times, so I'd also like to understand why it was changed from the approach you've written here last time. I'll try and look back when I get chance, but it would be great if you could review older revisions/PRs to this part of the code to try to understand this also please

simondeziel reviewed (approved):
LGTM, thanks!

foster999 commented:
Thanks for adding those bits and taking a look back. I think you're right - the old method just added them regardless of use, so the method being replaced here was likely a solution to that

As others have confirmed the issue is resolved, I'll get it merged and released once everything is passing

simondeziel commented on pylint.cfg:
```diff
@@ -2,7 +2,7 @@
 extension-pkg-whitelist=lxml
 
 [MESSAGES CONTROL]
-disable=missing-docstring,similarities,fixme,bad-continuation
```
This same `bad-continuation` option also shows up in `.pre-commit-config.yaml`.

kartben replied:
yep, thanks. This was a half-baked push, didn't think anyone would catch it :)

kartben commented:
@foster999 I think we should be back on track.
Updated pre-commit.ci to fetch more recent versions of black, pylint, etc. since one of the errors was actually a hard crash due to a too old pylint version being pulled in and not being compatible with Python 3.12
Also dropped `bad-continuation` as per the warning that was showing up before the crash.
pre-commit run -a now passes locally

```
$ pre-commit  run -a
check json...............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
check-manifest...........................................................Passed
black....................................................................Passed
pylint...................................................................Passed
```

simondeziel commented:
@foster999 if/when you get a chance to take another look that'd be great as the warning is polluting our build logs making it harder to spot any real issue. Thanks!

foster999 commented:
Sorry for the huge delay getting round to this! Now included in version 3.4.6

kartben commented:
> Sorry for the huge delay getting round to this! Now included in version 3.4.6

Awesome, thanks! Do you have a sense of when this will be available through pypi? Thanks!

foster999 commented:
Deployment failed for that one, but it's now up on PyPI as 3.4.7 ☺️

simondeziel commented:
> Sorry for the huge delay getting round to this! Now included in version 3.4.6

@foster999, no worries at all, thanks for your work as maintainer and also thanks again @kartben for the PR!
